File: JY/10월 4주차/[1차]캐시.java
import java.util.*;

class Solution {
    public int solution(int cacheSize, String[] cities) {
        LinkedList<String> cache = new LinkedList<>();
        int answer = 0;

        if (cacheSize == 0) {
            return cities.length * 5;
        }

        for (String city : cities) {
            city = city.toUpperCase();

            if (cache.contains(city)) {
                // 캐시 히트의 경우
                cache.remove(city);
                cache.add(city);
                answer += 1;
            } else {
                // 캐시 미스의 경우
                if (cache.size() < cacheSize) {
                    cache.removeFirst(); // LRU 원소 제거
                }
                cache.add(city);
                answer += 5;
            }
        }

        return answer;
    }
}

File: BW/10월4주차/[1차]캐시.js
function isValid(cacheSize, city, LRU) {
    var flag = false;

    for (var i = 0; i < LRU.length; i++) {
        if (city === LRU[i]) {
            LRU.splice(i, 1);  // 해당 도시 제거
            LRU.push(city);    // 최신으로 갱신
            flag = true;       // hit 발생
            break;
        }
    }
    return flag;
}

function solution(cacheSize, cities) {
    var answer = 0;
    var LRU = [];
    var runningTime = 0;
    cities = cities.map(city => city.toLowerCase());

    if (cacheSize === 0) {  // 캐시 크기가 0일 경우
        return cities.length * 5;
    }

    for (var i = 0; i < cities.length; i++) {
        var city = cities[i];

        if (isValid(cacheSize, city, LRU)) {
            runningTime += 1;
        } else {
            if (LRU.length === cacheSize) {
                LRU.shift();  // LRU배열에 따라 가장 오래된 항목 제거
            }
            LRU.push(city);  // 새로운 도시 추가
            runningTime += 5;
        }
    }

    return runningTime;
}
